### ISSUE TITLE
HCL Parser ignores errors

### ISSUE THREAD

**Issue by leventyalcin** [hashicorp/terraform#7948]
Hi everyone, 

I'm not sure how I'm gonna tell this but I think there is some regression on HCL parser. It ignores some errors. 

I've tried with 0.6.16, 0.7.0-rc2 and 0.7.0, but the result is same all the time. 

Let me show you with an example; 

```
 leoman  /tmp/a  1  terraform -version
Terraform v0.7.0

 leoman  /tmp/a  cat m.tf
variale "t1" { default = "levent" }
variale "t2" { default = "t2" }
variale "t3" { default = "t3" }

output "test" {
  value = "${join(",", list( var.t1, var.t2, var.t3))}"
}
 leoman  /tmp/a  terraform apply
Error configuring: 3 error(s) occurred:

* output.test: missing dependency: var.t1
* output.test: missing dependency: var.t2
* output.test: missing dependency: var.t3
```

it didn't complain about typo with `variable/variale`.

When I do some typo with `output` it shows all the errors. 

```
 leoman  /tmp/a  1  sed -i.bak 's/output/outut/g' m.tf
 leoman  /tmp/a  cat m.tf
variale "t1" { default = "levent" }
variale "t2" { default = "t2" }
variale "t3" { default = "t3" }

outut "test" {
  value = "${join(",", list( var.t1, var.t2, var.t3))}"
}
 leoman  /tmp/a  terraform apply
There are warnings and/or errors related to your configuration. Please
fix these before continuing.

Errors:

  * 1 error(s) occurred:

* module root: 4 error(s) occurred:

* Unknown root level key: variale
* Unknown root level key: variale
* Unknown root level key: variale
* Unknown root level key: outut
 leoman  /tmp/a  1 
```

Cheers, 
Levent.

**Comment by jbardin** [comment]
Hi @leventyalcin, 

Thanks for the bug report, and the example!

This is somewhat confusing. The problem is in the order of operations, where the missing variables are reported before the `Unknown root level key:` errors. If you provide the missing variables, you will get:

```
* module root: 3 error(s) occurred:

* Unknown root level key: variale
* Unknown root level key: variale
* Unknown root level key: variale
```

I'll see if there's something we can do to report the syntax errors earlier in this case.

**Comment by leventyalcin** [comment]
Hi, 

I think it's the very same reason, one of my colleagues spent about an hour to find out an error which was a typo of `backend`/`backends` in `terraform {}` block.

Most importantly, I was trying to attach AWS managed policy as below

```
resource "aws_iam_policy_attachment" "jenkins" {
  name       = "foo-instance"
  roles      = ["${aws_iam_role.role.name}"]
  policy_arn = "arn:aws:iam::aws:policy/AdministratorAccess"
}
```

Normally `roles` argument must be `role` but mistakenly I've run the code as above. The end result was Administrator policy has been removed from all the groups/roles etc.

I think, now, it's getting really serious. 

Thanks, 
Levent.

**Comment by apparentlymart** [comment]
Hi @leventyalcin! Thanks for sharing that additional example.

We are currently working on a more robust configuration language parser that should address problems like this. We need to roll it out carefully because it will introduce some minor behavioral differences compared to the current parser. We'll have more info on this soon. The configuration language is the main user-interface for Terraform, so we take the usability of it seriously and will be investing a lot in its robustness over the next few major releases.

**Comment by leventyalcin** [comment]
Quick update; 

The second issue was not `roles`/`role` problem. Reading the documentation on mobile wasn't the best idea. `roles` is the correct one. Turned out I was using policy_attachment with a wrong way and Policy attachment is a global action and should've used aws_iam_role_policy instead. 

Only the first issue is valid.

Sorry for the noise. 

Thanks.

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence.

I just tried the following configuration on Terraform v0.12-alpha1:

```hcl
variale "t1" {
  default = "levent"
}
variale "t2" {
  default = "t2"
}
variale "t3" {
  default = "t3"
}

output "test" {
  value = "${join(",", list(var.t1, var.t2, var.t3))}"
}
```

As expected, I got the following error messages about it:

![Error: Unsupported Block Type](https://user-images.githubusercontent.com/20180/47587389-8eea6000-d917-11e8-9ee2-1059803e14f9.png)

Terraform v0.12 in general now uses a schema-based decoding strategy that allows it to reliably detect when unexpected blocks and arguments are present. Validation oWitf references in expressions (like `var.t1`, etc) is also now deferred until after the configuration is loaded and its structure validated, so the references to undeclared variables are no longer presented in preference to the incorrect block name error.

With all of this said, it looks like this issue is now fixed in v0.12.0-alpha1 and so the fix will be included in the subsequent v0.12.0 final release. Thanks for reporting this issue, and sorry for the long wait for a fix!
